[SDK] enable ip-discovery

MythreyaK opened the pull request:
## Motivation

Update agent registration to use `aqlprofile_register_agent_info()` to register agent, to enable IP discovery with the SDK. 

## Technical Details

Uses the new API, [added here](https://github.com/ROCm/rocm-systems/blob/7c4369bde4e7843a5416678c9b4602a61a066d02/projects/aqlprofile/src/core/include/aql_profile_v2.h#L170-L180)

## Test Plan

> Any GC counter test should cover this. [that the new IP discovery mechanism still works with existing counters]
> @bing-ma 

`rocprofv3-test-counter-collection-txt-pmc2-execute` uses `GRBM_COUNT`

## Test Result

Counter collection tests should pass. 

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

MythreyaK committed:
[SDK] enable ip-discovery

ammarwa commented:
<details>
<summary><h2>Code Coverage Report</h2></summary>

<!-- Keep the empty line above to render markdown properly -->
# Code Coverage Report

## Tests Only

![code coverage tests.png](https://github.com/ROCm/rocm-systems/blob/ff45f8211fa889c87714897d2e8925f129c69f02/pr-368/tests.png?raw=True)

## Samples Only

![code coverage samples.png](https://github.com/ROCm/rocm-systems/blob/ff45f8211fa889c87714897d2e8925f129c69f02/pr-368/samples.png?raw=True)

## Tests + Samples

![code coverage all.png](https://github.com/ROCm/rocm-systems/blob/ff45f8211fa889c87714897d2e8925f129c69f02/pr-368/all.png?raw=True)

<!-- code-coverage-comment-identifier -->
</details>

bing-ma commented on projects/aqlprofile/src/core/pm4_factory.cpp:
```diff
@@ -52,15 +53,17 @@ locked_agent_cache& get_cache() {
   int_agent_info.domain = agent_info->domain;
   int_agent_info.bdf_id = agent_info->location_id;
 
+  std::strncpy(int_agent_info.name, int_agent_info.gfxip, 63);
```
Should use _sizeof(int_agent_info.name) - 1_ instead of 63.

MythreyaK replied:
Added a `static_assert` + `sizeof(int_agent_info.name) - 1`

MythreyaK replied:
Resolved in a different PR

bing-ma reviewed (dismissed):
Made a comment for a small change. Otherwise looks good to me.

MythreyaK committed:
code review

bing-ma reviewed (commented):
Is the **Motivation** statement auto-generated?
"Use V1 aqlprofile API for agent registration" is an inaccurate statement. You are using a new API aqlprofile_register_agent_info() to register agent, not a "V1 aqlprofile API".

MythreyaK commented:
> Is the Motivation statement auto-generated?
"Use V1 aqlprofile API for agent registration" is an inaccurate statement. You are using a new API aqlprofile_register_agent_info() to register agent, not a "V1 aqlprofile API".

Ah my bad, I was conflated the API and the version enumeration. Fixed!

bwelton commented on projects/aqlprofile/src/core/pm4_factory.cpp:
```diff
@@ -52,15 +53,18 @@ locked_agent_cache& get_cache() {
   int_agent_info.bdf_id = agent_info->location_id;
 
+  static_assert(sizeof(int_agent_info.name) == 64);
+  std::strncpy(int_agent_info.name, agent_info->agent_gfxip, sizeof(int_agent_info.name) - 1);
```
This should be ok since {} init's name to null.

MythreyaK replied:
Fixed in #687

bwelton commented on projects/aqlprofile/src/core/pm4_factory.cpp:
```diff
@@ -52,15 +53,18 @@ locked_agent_cache& get_cache() {
+  std::strncpy(int_agent_info.name, agent_info->agent_gfxip, sizeof(int_agent_info.name) - 1);
+
   auto len = strlen(agent_info->agent_gfxip);
   memset(int_agent_info.gfxip, 0, sizeof(int_agent_info.gfxip));
```
why are we copying the same string to two different fields? Can you open a ticket to unify these.

MythreyaK replied:
Done

bing-ma replied:
Internal _AgentInfo_ uses _name_ and _gfxip_ differently. For aqlprofile_agent_info, there is no _name_ member and _gfxip_ is the equivalent of the _AgentInfo.name_. Removing _name_ from _AgentInfo_ will cause a lot of code changes in the old aqlprofile code patch.

bwelton replied:
This is why i suggested opening a ticket here. May not be feasible to eliminate the duplication.

bing-ma replied:
OK, I see. So for "unify these", you mean unify legacy and v2, not unify name and gfxip.

Mythreya, can you check if you need to open a ticket or you can achieve the goal within this ticket?

MythreyaK replied:
Created an internal ticket for this. I'll probably work on this after this PR gets merged, because of the dependency on counters.

jrmadsen reviewed (changes requested):
Where is the versioning change to AQLProfile? Can we handle older versions of AQLProfile at compile time? Runtime?

MythreyaK commented:
> Where is the versioning change to AQLProfile? Can we handle older versions of AQLProfile at compile time? Runtime?

Ah, I missed that. The header from aqlprofile project does not seem to have a version macro. I'll open a PR for that.

MythreyaK commented:
Created #706 for versioning

ApoKalipse-V commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -939,14 +939,18 @@ get_aql_handles()
-                    .shader_arrays_per_se = agent->simd_arrays_per_engine};
+                    .shader_arrays_per_se = agent->simd_arrays_per_engine,
+                    .domain               = agent->domain,
+                    .location_id          = agent->location_id,
```
Since we are adding more fields to agent today, can we add the realtime frequency?
ATT uses hsa runtime to query these, but if we want to get ATT to use aqlprofile_agent, we will need this information.
@bing-ma  I think SPM also has a few more parameters no?

bing-ma replied:
@ApoKalipse-V, SPM is fine because it only needs XCC/SE/SA which are already available.

MythreyaK replied:
@ApoKalipse-V I am not sure of what the relevant changes are. Is it [one of these](https://github.com/ROCm/rocm-systems/blob/develop/projects/rocprofiler-sdk/source/include/rocprofiler-sdk/agent.h#L181-L183)?

<img width="1263" height="229" alt="image" src="https://github.com/user-attachments/assets/13ab1464-ba9d-46b1-a774-a4e980b1b126" />

MythreyaK replied:
Will create a ticket for this.

bing-ma reviewed (approved):
Looks good to me after the revisions.

MythreyaK committed:
Update IP discovery in sdk to use aql versioning

Use version macros to conditionally register agents

MythreyaK commented:
CI failure is due to RCCL issue, fix in #985

bing-ma commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -939,18 +977,7 @@ get_aql_handles()
-                    ROCP_WARNING << "Failed to register agent " << agent->name;
-                }
-                agent_handles.push_back(handle);
+                agent_handles.push_back(register_aql_agent(agent));
```
Using function name register_aql_agent() gives me a very strange feel. Maybe call it get_aqlprofile_agent_handle().

MythreyaK replied:
Done

bing-ma reviewed (approved):
Added a comment for a minor change. Otherwise looks good to me.

MythreyaK committed:
code review

MythreyaK committed:
code review

bing-ma commented on projects/aqlprofile/src/core/pm4_factory.cpp:
```diff
@@ -53,6 +53,11 @@ locked_agent_cache& get_cache() {
+  // If assertion fails,
+  // Update handling of new fields
+  // Update size to new size
+  static_assert(sizeof(aqlprofile_agent_info_v1_t) == 40);
```
Is this how we usually handle the 'size' field? I thought we would compare the data member offset with size to make sure we can access this member. No assertion is needed.

MythreyaK replied:
I added this just as a way to ensure that the new field and size is correctly handled in the updated code (in the checks that follow) and not forgotten. 

> I thought we would compare the data member offset with size to make sure we can access this member.

Since we have only one size, I thought it was okay to exclude the check. Updated with a check to make it explicit.

MythreyaK committed:
code review

bwelton commented on projects/aqlprofile/src/core/pm4_factory.cpp:
```diff
@@ -52,14 +55,27 @@ locked_agent_cache& get_cache() {
+
+  static_assert(sizeof(int_agent_info.name) == 64);
 
   auto len = strlen(agent_info->agent_gfxip);
```
We need a check here for null.

MythreyaK replied:
Done

bwelton commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/aql/aql_profile_v2.h:
```diff
@@ -44,6 +46,58 @@ typedef enum
+    AQLPROFILE_BLOCK_NAME_RESERVED_9,
+
+    // Add new blocks above
+    AQLPROFILE_BLOCKS_NUMBER
```
Use AQLPROFILE_BLOCK_LAST

MythreyaK replied:
This is already present in aqlprofile codebase; this change is copied from there for rocprofiler-sdk's local copy. Should I update both headers with this change?

bwelton reviewed (approved):
Approve, make the small changes before landing.

jrmadsen commented on projects/aqlprofile/src/core/logger.h:
```diff
@@ -40,9 +40,11 @@
 namespace aql_profile {
 
-class Logger {
+class DEFAULT_VISIBILITY Logger {
```
Why is this class exported?

MythreyaK replied:
target 'pm4-factory-test' fails to find definitions for `aql_profile::Logger` because they are not visible to the test that links to the aqlprofile shared library.

I wasn't sure how to cleanly let the test find these symbols other than making them exported symbols.

We don't install these tests as far as I can tell, so is static linking to the aqlprofile library for this test an acceptable alternative?

jrmadsen replied:
Yes, building a static library (that we don't install) is how we work around this issue in rocprofiler-sdk

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/aql/aql_profile_v2.h:
```diff
@@ -184,18 +291,27 @@ aqlprofile_get_pmc_info(const aqlprofile_pmc_profile_t* profile,
                         aqlprofile_pmc_info_type_t      attribute,
                         void*                           value);
 
+typedef enum aqlprofile_att_parameter_rt_timestamp_t
```
What is `rt` short for? Abbreviations, in general, should be avoided.

MythreyaK replied:
This change is already present in the header from aqlprofile, so this change is just updating rocprofiler-sdk's copy of the header. Do I update this in both files, or should that be a separate PR?

bing-ma replied:
@ApoKalipse-V, Giovanni, why is this not already defined in sdk's copy? And can we not use abbreviation for this case?

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/agent.cpp:
```diff
@@ -930,6 +930,50 @@ get_agent_mapping()
     static auto*& _v = common::static_object<std::vector<agent_pair>>::construct();
     return *CHECK_NOTNULL(_v);
 }
+aqlprofile_agent_handle_t
```
```suggestion

aqlprofile_agent_handle_t
```

MythreyaK replied:
Sorry is the suggestion to add a new line here? The change suggestion here renders as just adding a new line before `aqlprofile_agent_handle_t`

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/aql/version.h:
Why does rocprofiler-sdk have a version.h for AQLProfile?

MythreyaK replied:
My understanding is that the header for `aqlprofile` is from `rocprofiler-sdk`'s local copy, but the library is found with `find_library`. So, to get the versioning macros, we need to add a copy of that header as well alongside aqlprofile's API header. Will update this with any alternate approaches you suggest.

bing-ma replied:
I always wondered why it is designed in this way, maybe because of separated repos before monorepo? Either Ben or Giovanni should know the definitive answer. With monorepo, can we add the aqlprofile header path directly into SDK's cmake files?

MythreyaK replied:
> can we add the aqlprofile header path directly into SDK's cmake files?

That would probably break builds in the [standalone repo](https://github.com/ROCm/rocprofiler-sdk), which from my understanding, is used for other release activities (e.g., [7.0 release](https://github.com/ROCm/rocprofiler-sdk/tree/release/rocm-rel-7.0)).

MythreyaK committed:
Merge rocprofiler-sdk copy of aql version and api header

MythreyaK committed:
aqlprofile: add object library and simplify tests linkage

MythreyaK committed:
[aqlprofile] update `aqlprofile_agent_info_v1_t`

Update `aqlprofile_agent_info_v1_t` with size field, bump
minor version. Handle the new ABI change in rocprofiler-sdk

bing-ma reviewed (approved):
Looks good to me.

bwelton commented on projects/aqlprofile/src/core/pm4_factory.cpp:
```diff
@@ -52,14 +56,34 @@ locked_agent_cache& get_cache() {
+    int_agent_info.bdf_id = agent_info->location_id;
+  } else {
+    ERR_LOGGING << "Unsupported agent_info size " << agent_info->size;
+    throw std::runtime_error("Unsupported agent_info size");
```
We do need to get rid of these throws. This is not for this PR but we should have a more dedicated error pathing.

bwelton commented on projects/aqlprofile/src/core/pm4_factory.cpp:
```diff
@@ -52,14 +56,34 @@ locked_agent_cache& get_cache() {
+    throw std::runtime_error("Unsupported agent_info size");
+  }
+
+  static_assert(sizeof(int_agent_info.name) == 64);
```
Move to after the allocation on 65. Good check :-)

bwelton commented on projects/aqlprofile/src/core/pm4_factory.cpp:
```diff
@@ -52,14 +56,34 @@ locked_agent_cache& get_cache() {
+  aqlprofile_agent_handle_t agent_id{};
+  AgentInfo int_agent_info{};
+
+  if (agent_info->size == 40) {
```
Id invert this if statement to reduce indentation. 


```suggestion
  if (agent_info->size != 40) {
      //error
  }
  
  // copy data.
```

bwelton reviewed (approved):
Mostly minor nits, thanks for the change!

MythreyaK committed:
code review

bgopesh reviewed (changes requested):
I am okay with the changes but I don't see any tests running and there are still requested changes from Jonathan. Please clear them, if already addressed

jrmadsen commented on projects/aqlprofile/src/core/include/aqlprofile-sdk/aql_profile_v2.h:
```diff
@@ -197,6 +197,7 @@ typedef struct {
  * (if loaded) or the KFD topology.
  */
 typedef struct {
+  uint64_t size;           /**< Size of this structure */
```
Yeah... we need a `aqlprofile_agent_info_v2_t` here.

jrmadsen replied:
IDK how we will be able to handle this ABI-wise.

MythreyaK replied:
Had a short discussion with @bwelton a little while back. If I'm not mistaken, we have 3 cases, 

* For aqlprofile library loaded at runtime before the versioning change, the symbols that check the version won't be found, so we won't have issues with corrupted ABI (the application will fail to run).
* If at runtime, we detect version aqlprofile 1.0.0, the old structure is used from rocprofiler-sdk (the ` aqlprofile_agent_info_v1_t_100` structure in agent.cpp). 
* If the version is 1.1.0, we use the new one. 

If we go with a new v2 structure, the same 3 variations remain. 
* Pre-versioning support, application will not run because of missing symbols in aqlprofile. 
* for 1.0.0, we use the v1 structure + API function 
* for 1.1.0, we use the v2 structure + API function 

Do I add a new structure v2 instead of updating the existing one?

jrmadsen replied:
> Pre-versioning support, application will not run because of missing symbols in aqlprofile.

This is what I am worried about. I would prefer to not be tied to a specific version of AQLProfile.

MythreyaK replied:
Added a fallback with a weak symbol for versioning as discussed.

MythreyaK committed:
Provide fallback for old aqlprofile

MythreyaK commented:
Closed in favor of #3626.
